src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Setup/SetupComponentRegistrationService.cs

```diff
@@ -0,0 +1,259 @@
+    private readonly Lazy<HashSet<string>?> _installedRuntimeComponentIds;
+
+    private IVsSetupCompositionService? _setupCompositionService;
+    private IMissingComponentRegistrationService? _missingComponentRegistrationService;
+    private IAsyncDisposable? _solutionEventsSubscription;
+    private ImmutableDictionary<Guid, UnconfiguredSetupComponentSnapshot?> _snapshotByProjectGuid = ImmutableDictionary<Guid, UnconfiguredSetupComponentSnapshot?>.Empty;
+
+    [ImportingConstructor]
+    public SetupComponentRegistrationService(
+        IVsService<SVsBrokeredServiceContainer, IBrokeredServiceContainer> serviceBrokerContainer,
+        IVsService<SVsSetupCompositionService, IVsSetupCompositionService> vsSetupCompositionService,
+        ISolutionService solutionService,
+        IProjectFaultHandlerService projectFaultHandlerService,
+        JoinableTaskContext joinableTaskContext)
+        : base(new(joinableTaskContext))
+    {
+        _serviceBrokerContainer = serviceBrokerContainer;
+        _setupCompositionServiceFactory = vsSetupCompositionService;
+        _solutionService = solutionService;
+        _projectFaultHandlerService = projectFaultHandlerService;
+
+        _installedRuntimeComponentIds = new Lazy<HashSet<string>?>(FindInstalledRuntimeComponentIds);
+
+        static HashSet<string>? FindInstalledRuntimeComponentIds()
+        {
+            // Workaround for https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1460328
+            // VS Setup doesn't know about runtimes installed outside of VS. Deep detection is not suggested for performance reasons.
+            // Instead we read installed runtimes from the registry.
+
+            // TODO Consider the architecture of the project itself.
+            // We currently assume that the project will run on the same architecture as VS.
+            // This will be the common case, but does not cover situations where a project runs
+            // under emulation (e.g. an x64 build running on an ARM64 system).
+            Architecture architecture = RuntimeInformation.ProcessArchitecture;
+
+            string[]? runtimeVersions = NetCoreRuntimeVersionsRegistryReader.ReadRuntimeVersionsInstalledInLocalMachine(architecture);
+
+            if (runtimeVersions is null)
+            {
+                return null;
+            }
+
+            HashSet<string> installedRuntimeComponentIds = new(StringComparers.VisualStudioSetupComponentIds);
+
+            foreach (string runtimeVersion in runtimeVersions)
+            {
+                // Example versions:
+                //
+                // - "3.1.32"
+                // - "7.0.11"
+                // - "8.0.0-preview.7.23375.6"
+                // - "8.0.0-rc.1.23419.4"
+                //
+                // We only want the major/minor version, in the same format as TargetFrameworkVersion (e.g. "v8.0").
+
+                int firstDotIndex = runtimeVersion.IndexOf('.');
```

Comment:
This might be written as a loop through the characters, which would avoid the need for the substring, at the cost of being as readable

Reply:
Can you explain how a loop would help? `IndexOf` is conceptually a loop (though it is able to use SIMD to parallelise the work, which it does on .NET Core but possibly not on .NET Framework).

I don't think we could reduce allocations here though without changing `ComponentIdByRuntimeVersion` to be keyed on values without the `v` prefix.

This code runs once for the whole solution at startup, and the loop is likely to iterate 2-10 times or so, depending on how many versions of the runtime you have installed.